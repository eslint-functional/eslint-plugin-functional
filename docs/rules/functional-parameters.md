# Enforce Functional Parameters (functional-parameters)

Disallow use of rest parameters, the `arguments` keyword and enforces that functions take a least 1 parameter.

## Rule Details

In functions, `arguments` is a special variable that is implicitly available.
This variable is an array containing the arguments passed to the function call and is often used to allow for any number of parameters to be passed to a function. Rest parameters are another way any number of parameters can be passed to a function.

When it comes to functional programming, known and explicit parameters must be used.

Note: With an unknown number of parameters, currying functions is a lot more difficult/impossible.

Examples of **incorrect** code for this rule:

```js
/* eslint functional/functional-parameters: "error" */

function add() {
  return arguments.reduce((sum, number) => sum + number, 0);
}
```

```js
/* eslint functional/functional-parameters: "error" */

function add(...numbers) {
  return numbers.reduce((sum, number) => sum + number, 0);
}
```

Examples of **correct** code for this rule:

```js
/* eslint functional/functional-parameters: "error" */

function add(numbers) {
  return numbers.reduce((sum, number) => sum + number, 0);
}
```

## Options

This rule accepts an options object of the following type:

```ts
{
  allowRestParameter: boolean;
  allowArgumentsKeyword: boolean;
  enforceParameterCount: false | "atLeastOne" | "exactlyOne" | {
    count: "atLeastOne" | "exactlyOne";
    ignoreIIFE: boolean;
  };
  ignorePattern?: string | Array<string>;
}
```

The default options:

```ts
{
  allowRestParameter: false,
  allowArgumentsKeyword: false,
  enforceParameterCount: {
    count: "atLeastOne",
    ignoreIIFE: true
  }
}
```

Note: the `lite` ruleset overrides the default options to:

```ts
{
  allowRestParameter: false,
  allowArgumentsKeyword: false,
  enforceParameterCount: false
}
```

### `allowRestParameter`

If true, this option allows for the use of rest parameters.

### `allowArgumentsKeyword`

If true, this option allows for the use of the `arguments` keyword.

### `enforceParameterCount`

This option enforces the number of parameters each function takes.

#### false

Don't enforce a parameter count.

#### "atLeastOne"

Require all functions to have a least one parameter.

There's not much point of having a function that doesn't take any parameters in functional programming.

#### "exactlyOne"

Require all functions to have exactly one parameter.

Any function that take takes multiple parameter can be rewritten as a higher-order function that only takes one.

Example:

```js
// This function
function add(x, y) {
  return x + y;
}

// can be rewritten like this.
function add(x) {
  return (y) => x + y;
}
```

See [Currying](https://en.wikipedia.org/wiki/Currying) and [Higher-order function](https://en.wikipedia.org/wiki/Higher-order_function) on Wikipedia for more infomation.

#### `enforceParameterCount.count`

See [enforceParameterCount](#enforceparametercount).

#### `enforceParameterCount.ignoreIIFE`

If true, this option allows for the use of [IIFEs](https://developer.mozilla.org/en-US/docs/Glossary/IIFE) that do not have any parameters.

### `ignorePattern`

Patterns will be matched against function names.
See the [ignorePattern](./options/ignore-pattern.md) docs for more infomation.
