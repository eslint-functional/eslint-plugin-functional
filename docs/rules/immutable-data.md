# Disallow mutating objects and arrays (immutable-data)

This rule prohibits syntax that mutates existing objects and arrays via assignment to or deletion of their properties/elements.

## Rule Details

While requiring the `readonly` modifier forces declared types to be immutable,
it won't stop assignment into or modification of untyped objects or external types declared under different rules.

Examples of **incorrect** code for this rule:

```js
/* eslint functional/immutable-data: "error" */

const obj = { foo: 1 };

obj.foo += 2; // <- Modifying an existing object/array is not allowed.
obj.bar = 1; // <- Modifying an existing object/array is not allowed.
delete obj.foo; // <- Modifying an existing object/array is not allowed.
Object.assign(obj, { bar: 2 }); // <- Modifying properties of existing object not allowed.
```

```js
/* eslint functional/immutable-data: "error" */

const arr = [0, 1, 2];

arr[0] = 4; // <- Modifying an array is not allowed.
arr.length = 1; // <- Modifying an array is not allowed.
arr.push(3); // <- Modifying an array is not allowed.
```

Examples of **correct** code for this rule:

```js
/* eslint functional/immutable-data: "error" */

const obj = { foo: 1 };
const arr = [0, 1, 2];

const x = {
  ...obj
  bar: [
    ...arr, 3, 4
  ]
}
```

## Options

This rule accepts an options object of the following type:

```ts
{
  assumeTypes:
    | boolean
    | {
        forArrays: boolean;
        forObjects: boolean;
      }
  ignoreImmediateMutation: boolean;
  ignorePattern?: string | Array<string>;
  ignoreAccessorPattern?: string | Array<string>;
}
```

The default options:

```ts
{
  assumeTypes: true,
  ignoreImmediateMutation: true,
};
```

### `assumeTypes`

The rule take advantage of TypeScript's typing engine to check if mutation is taking place.
If you are not using TypeScript, type checking cannot be performed; hence this option exists.

This option will make the rule assume the type of the nodes it is checking are of type Array/Object.  
However this may result in some false positives being picked up.

Disabling this option can result in false negatives, for example:

```js
// When this option is DISABLED (and type info is not available).
const x = [0, 1, 2];
x.push(3); // This will NOT be flagged.
// This is due to the fact that without a typing engine, we cannot tell that x is an array.
```

Note: This option will have no effect if the TypeScript typing engine is avaliable (i.e. you are using TypeScript and have configured eslint correctly).

### `ignoreImmediateMutation`

If true, immediate mutation of objects before they are assigned to a variable is allowed.
This allows for the use of array mutator methods to be chained to newly created arrays.

For example, an array can be immutably sorted like so:

```js
const original = ["foo", "bar", "baz"];
const sorted = [...original].sort((a, b) => a.localeCompare(b)); // This is OK with ignoreImmediateMutation.
```

### `ignorePattern`

Patterns will be matched against variable names.
See the [ignorePattern](./options/ignore-pattern.md) docs for more infomation.

### `ignoreAccessorPattern`

This option takes a match string or an array of match strings (not a RegExp pattern).

The match string allows you to specify dot seperated `.` object paths and has support for "glob" `*` and "globstar" `**` matching.

For example:

```js
{
  // Ignore all reassigning to object properties that are prefixed with "mutable_".
  "ignoreAccessorPattern": "**.mutable_*"
}
```

```js
{
  // Ignore all shallow mutations made to object properties that are prefixed with "mutable_".
  "ignoreAccessorPattern": "**.mutable_*.*"
}
```

```js
{
  // Ignore all deep mutations made to object properties that are prefixed with "mutable_".
  "ignoreAccessorPattern": "**.mutable_*.*.**"
}
```

```js
{
  // Ignore all deep mutations and reassigning to object properties that are prefixed with "mutable_".
  "ignoreAccessorPattern": "**.mutable_*.**"
  // This is the same as `"ignoreAccessorPattern": ["**.mutable_*", "**.mutable_*.*.**"]`
}
```

#### Wildcards

The following wildcards can be used when specifing a pattern:

`**` - Match any depth (including zero). Can only be used as a full accessor.
`*` - When used as a full accessor, match the next accessor (there must be one). When used as part of an accessor, match any characters.
