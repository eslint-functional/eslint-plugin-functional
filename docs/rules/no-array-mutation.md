### no-array-mutation

[![Type Info Required][type-info-badge]][type-info-url]

This rule prohibits mutating an array via assignment to or deletion of their elements/properties. This rule enforces array immutability without the use of `ReadonlyArray<T>` (as apposed to [readonly-array](#readonly-array)).

```typescript
const x = [0, 1, 2];

x[0] = 4; // <- Mutating an array is not allowed.
x.length = 1; // <- Mutating an array is not allowed.
x.push(3); // <- Mutating an array is not allowed.
```

#### Has Fixer

No

#### Options

- [ignore-prefix](#using-the-ignore-prefix-option)
- [ignore-suffix](#using-the-ignore-suffix-option)
- [ignore-pattern](#using-the-ignore-pattern-option)
- [ignore-new-array](#using-the-ignore-new-array-option-with-no-array-mutation)
- ~~ignore-mutation-following-accessor~~ - _deprecated in favor of [ignore-new-array](#using-the-ignore-new-array-option-with-no-array-mutation)_

#### Example config

```javascript
"no-array-mutation": true
```

```javascript
"no-array-mutation": [true, {"ignore-prefix": "mutable"}]
```

```javascript
"no-array-mutation": [true, "ignore-new-array"]
```
