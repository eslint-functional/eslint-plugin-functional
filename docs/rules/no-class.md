# Disallow classes (no-class)

Disallow use of the `class` keyword.

## Rule Details

Thanks to libraries like [recompose](https://github.com/acdlite/recompose) and Redux's [React Container components](http://redux.js.org/docs/basics/UsageWithReact.html), there's not much reason to build Components using `React.createClass` or ES6 classes anymore. The `no-this` rule makes this explicit.

```typescript
const Message = React.createClass({
  render: function() {
    return <div>{this.props.message}</div>; // <- no this allowed
  }
});
```

Instead of creating classes, you should use React 0.14's [Stateless Functional Components](https://medium.com/@joshblack/stateless-components-in-react-0-14-f9798f8b992d#.t5z2fdit6) and save yourself some keystrokes:

```typescript
const Message = ({ message }) => <div>{message}</div>;
```

What about lifecycle methods like `shouldComponentUpdate`? We can use the [recompose](https://github.com/acdlite/recompose) library to apply these optimizations to your Stateless Functional Components. The [recompose](https://github.com/acdlite/recompose) library relies on the fact that your Redux state is immutable to efficiently implement shouldComponentUpdate for you.

```typescript
import { pure, onlyUpdateForKeys } from "recompose";

const Message = ({ message }) => <div>{message}</div>;

// Optimized version of same component, using shallow comparison of props
// Same effect as React's PureRenderMixin
const OptimizedMessage = pure(Message);

// Even more optimized: only updates if specific prop keys have changed
const HyperOptimizedMessage = onlyUpdateForKeys(["message"], Message);
```

## Options

The rule does not accept any options.
