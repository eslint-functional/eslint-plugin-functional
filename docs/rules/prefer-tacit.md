# Tacit / Point-Free (prefer-tacit)

This rule enforces using functions directly if they can be without wrapping them.

## Rule Details

Generally there's no reason to wrap a function with a callback wrapper if it's directly called anyway.
Doing so creates extra inline lambdas that slow the runtime down.

Examples of **incorrect** code for this rule:

```ts
/* eslint functional/prefer-tacit: "error" */

function f(x) {
  // ...
}

const foo = x => f(x);
```

Examples of **correct** code for this rule:

```ts
/* eslint functional/prefer-tacit: "error" */

function f(x) {
  // ...
}

const foo = f;
```

## Options

This rule accepts an options object of the following type:

```ts
{
  assumeTypes:
    | false
    | {
        allowFixer: boolean;
      }
  ignorePattern?: string | Array<string>;
}
```

The default options:

```ts
{
  assumeTypes: false,
};
```

### `assumeTypes`

The rule take advantage of TypeScript's typing engine to check if callback wrapper is in fact safe to remove.

This option will make the rule assume that the function only accepts the arguments given to it in the wrapper.  
However this may result in some false positives being picked up.

```js
const foo = x => f(x);    // If `f` only accepts one parameter then this is violation of the rule.
const bar = foo(1, 2, 3); // But if `f` accepts more than one parameter then it isn't.
```

Note: Enabling this option is the only way to get this rule to report violations in an environment without TypeScript's typing engine avaliable (e.g. In Vanilla JS).

### `assumeTypes.allowFixer`

When set types will be assumed.

This option states whether the auto fixer should be enabled for violations that are found when assuming types (violations found without assuming types will ignore this option).
Because when assuming types, false positives may be found, it's recommended to set this option to `false`.

### `ignorePattern`

See the [ignorePattern](./options/ignore-pattern.md) docs.
