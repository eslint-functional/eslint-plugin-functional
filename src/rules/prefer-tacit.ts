import type { TSESTree } from "@typescript-eslint/experimental-utils";
import type { ReportDescriptor } from "@typescript-eslint/experimental-utils/dist/ts-eslint";
import { all as deepMerge } from "deepmerge";
import type { JSONSchema4 } from "json-schema";
import type { FunctionLikeDeclaration, Type } from "typescript";

import type { IgnorePatternOption } from "~/common/ignore-options";
import { ignorePatternOptionSchema } from "~/common/ignore-options";
import type { RuleContext, RuleMetaData, RuleResult } from "~/utils/rule";
import { createRule, getESTreeNode, getTypeOfNode } from "~/utils/rule";
import {
  isBlockStatement,
  isCallExpression,
  isFunctionLike,
  isIdentifier,
  isReturnStatement,
  isTSFunctionType,
} from "~/utils/typeguard";

// The name of this rule.
export const name = "prefer-tacit" as const;

// The options this rule can take.
type Options = IgnorePatternOption & {
  readonly assumeTypes:
    | false
    | {
        readonly allowFixer: boolean;
      };
};

// The schema for the rule options.
const schema: JSONSchema4 = [
  deepMerge([
    ignorePatternOptionSchema,
    {
      type: "object",
      properties: {
        ignoreImmediateMutation: {
          type: "boolean",
        },
        assumeTypes: {
          oneOf: [
            {
              type: "boolean",
              enum: [false],
            },
            {
              type: "object",
              properties: {
                allowFixer: {
                  type: "boolean",
                },
              },
              additionalProperties: false,
            },
          ],
        },
      },
      additionalProperties: false,
    },
  ]),
];

// The default options for the rule.
const defaultOptions: Options = {
  assumeTypes: false,
};

// The possible error messages.
const errorMessages = {
  generic: "Potentially unnecessary function wrapper.",
} as const;

// The meta data for this rule.
const meta: RuleMetaData<keyof typeof errorMessages> = {
  type: "suggestion",
  docs: {
    description: "Replaces `x => f(x)` with just `f`.",
    category: "Best Practices",
    recommended: "warn",
  },
  messages: errorMessages,
  fixable: "code",
  schema,
};

/**
 * From the callee's type, does it follow that the caller violates this rule.
 */
function isCallerViolation(
  caller: TSESTree.CallExpression,
  calleeType: Type,
  context: RuleContext<keyof typeof errorMessages, Options>
): boolean {
  if (calleeType.symbol === undefined) {
    return false;
  }
  const tsDeclaration =
    calleeType.symbol.valueDeclaration ?? calleeType.symbol.declarations?.[0];

  if (tsDeclaration === undefined) {
    return false;
  }
  const declaration = getESTreeNode(tsDeclaration, context);

  return (
    (declaration !== null &&
      (isFunctionLike(declaration) || isTSFunctionType(declaration)) &&
      declaration.params.length === caller.arguments.length) ||
    // Check for optional params.
    (tsDeclaration as FunctionLikeDeclaration).parameters
      .slice(caller.arguments.length)
      .every(
        (param) =>
          param.initializer !== undefined || param.questionToken !== undefined
      )
  );
}

/**
 * Check for violations based on the given caller.
 */
function getCallDescriptors(
  node:
    | TSESTree.FunctionDeclaration
    | TSESTree.FunctionExpression
    | TSESTree.ArrowFunctionExpression,
  context: RuleContext<keyof typeof errorMessages, Options>,
  options: Options,
  caller: TSESTree.CallExpression
): Array<ReportDescriptor<keyof typeof errorMessages>> {
  if (
    isIdentifier(caller.callee) &&
    node.params.length === caller.arguments.length &&
    node.params.every((param, index) => {
      const callArg = caller.arguments[index];
      return (
        isIdentifier(callArg) &&
        isIdentifier(param) &&
        callArg.name === param.name
      );
    })
  ) {
    const calleeType = getTypeOfNode(caller.callee, context);
    const assumingTypes =
      (calleeType === null || calleeType.symbol === undefined) &&
      options.assumeTypes !== false;

    if (
      assumingTypes ||
      (calleeType !== null && isCallerViolation(caller, calleeType, context))
    ) {
      const calleeName = caller.callee.name;
      return [
        {
          node,
          messageId: "generic",
          fix:
            // No fixer when assuming types as this is dangerous.
            (typeof options.assumeTypes !== "object" && assumingTypes) ||
            // Unless user specifies they want it.
            (typeof options.assumeTypes === "object" &&
              !options.assumeTypes.allowFixer)
              ? undefined
              : (fixer) => fixer.replaceText(node, calleeName),
        },
      ];
    }
    return [];
  }
  return [];
}

/**
 * Check for violations in the form: `x => f(x)`.
 */
function getDirectCallDescriptors(
  node:
    | TSESTree.FunctionDeclaration
    | TSESTree.FunctionExpression
    | TSESTree.ArrowFunctionExpression,
  context: RuleContext<keyof typeof errorMessages, Options>,
  options: Options
): Array<ReportDescriptor<keyof typeof errorMessages>> {
  if (isCallExpression(node.body)) {
    return getCallDescriptors(node, context, options, node.body);
  }
  return [];
}

/**
 * Check for violations in the form: `x => { return f(x); }`.
 */
function getNestedCallDescriptors(
  node:
    | TSESTree.FunctionDeclaration
    | TSESTree.FunctionExpression
    | TSESTree.ArrowFunctionExpression,
  context: RuleContext<keyof typeof errorMessages, Options>,
  options: Options
): Array<ReportDescriptor<keyof typeof errorMessages>> {
  if (
    isBlockStatement(node.body) &&
    node.body.body.length === 1 &&
    isReturnStatement(node.body.body[0]) &&
    node.body.body[0].argument !== null &&
    isCallExpression(node.body.body[0].argument)
  ) {
    return getCallDescriptors(
      node,
      context,
      options,
      node.body.body[0].argument
    );
  }
  return [];
}

/**
 * Check if the given function node violates this rule.
 */
function checkFunction(
  node:
    | TSESTree.FunctionDeclaration
    | TSESTree.FunctionExpression
    | TSESTree.ArrowFunctionExpression,
  context: RuleContext<keyof typeof errorMessages, Options>,
  options: Options
): RuleResult<keyof typeof errorMessages, Options> {
  return {
    context,
    descriptors: [
      ...getDirectCallDescriptors(node, context, options),
      ...getNestedCallDescriptors(node, context, options),
    ],
  };
}

// Create the rule.
export const rule = createRule<keyof typeof errorMessages, Options>(
  name,
  meta,
  defaultOptions,
  {
    FunctionDeclaration: checkFunction,
    FunctionExpression: checkFunction,
    ArrowFunctionExpression: checkFunction,
  }
);
