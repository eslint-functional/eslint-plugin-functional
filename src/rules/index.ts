import * as functionalParameters from "./functional-parameters";
import * as immutableData from "./immutable-data";
import * as noClass from "./no-class";
import * as noConditionalStatement from "./no-conditional-statement";
import * as noExpressionStatement from "./no-expression-statement";
import * as noLet from "./no-let";
import * as noLoop from "./no-loop-statement";
import * as noMethodSignature from "./no-method-signature";
import * as noMixedType from "./no-mixed-type";
import * as noPromiseReject from "./no-promise-reject";
import * as noReturnVoid from "./no-return-void";
import * as noThisExpression from "./no-this-expression";
import * as noThrowStatement from "./no-throw-statement";
import * as noTryStatement from "./no-try-statement";
import * as preferReadonlyTypes from "./prefer-readonly-type";
import * as preferType from "./prefer-type-literal";

/**
 * All of the custom rules.
 */
export const rules = {
  [functionalParameters.name]: functionalParameters.rule,
  [immutableData.name]: immutableData.rule,
  [noClass.name]: noClass.rule,
  [noConditionalStatement.name]: noConditionalStatement.rule,
  [noExpressionStatement.name]: noExpressionStatement.rule,
  [noLet.name]: noLet.rule,
  [noLoop.name]: noLoop.rule,
  [noMethodSignature.name]: noMethodSignature.rule,
  [noMixedType.name]: noMixedType.rule,
  [noPromiseReject.name]: noPromiseReject.rule,
  [noReturnVoid.name]: noReturnVoid.rule,
  [noThisExpression.name]: noThisExpression.rule,
  [noThrowStatement.name]: noThrowStatement.rule,
  [noTryStatement.name]: noTryStatement.rule,
  [preferReadonlyTypes.name]: preferReadonlyTypes.rule,
  [preferType.name]: preferType.rule,
};
