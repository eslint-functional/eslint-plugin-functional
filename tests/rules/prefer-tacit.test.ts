/**
 * @file Tests for prefer-tacit.
 */

import dedent from "dedent";
import { RuleTester } from "eslint";

import { name, rule } from "../../src/rules/prefer-tacit";
import { es6, typescript } from "../helpers/configs";
import type { InvalidTestCase, ValidTestCase } from "../helpers/util";
import {
  describeTsOnly,
  processInvalidTestCase,
  processValidTestCase,
} from "../helpers/util";

// Valid test cases.
const es6Valid: ReadonlyArray<ValidTestCase> = [
  {
    // No typedef for `f` therefore no error (when not assuming types).
    code: `const foo = x => f(x);`,
    optionsSet: [[]],
  },
];

// Valid test cases.
const tsValid: ReadonlyArray<ValidTestCase> = [
  ...es6Valid,
  // FunctionDeclaration.
  {
    code: dedent`
      function f(x, y) {}
      const foo = x => f(x);`,
    optionsSet: [[]],
  },
  // FunctionExpression.
  {
    code: dedent`
      const f = function(x, y) {}
      const foo = x => f(x);`,
    optionsSet: [[]],
  },
  // ArrowFunction.
  {
    code: dedent`
      const f = (x, y) => {}
      const foo = x => f(x);`,
    optionsSet: [[]],
  },
  // TypeAlias.
  {
    code: dedent`
      type F = (x, y) => {};
      const f = undefined as unknown as F;
      const foo = x => f(x);`,
    optionsSet: [[]],
  },
];

// Valid test cases.
const es6Invalid: ReadonlyArray<InvalidTestCase> = [
  {
    // No typedef but assuming types.
    // No fixer.
    code: `const foo = x => f(x);`,
    optionsSet: [[{ assumeTypes: { allowFixer: false } }]],
    errors: [
      {
        messageId: "generic",
        type: "ArrowFunctionExpression",
        line: 1,
        column: 13,
      },
    ],
  },
  {
    // No typedef but assuming types.
    // With fixer.
    code: `const foo = x => f(x);`,
    optionsSet: [[{ assumeTypes: { allowFixer: true } }]],
    output: dedent`
      const foo = f;`,
    errors: [
      {
        messageId: "generic",
        type: "ArrowFunctionExpression",
        line: 1,
        column: 13,
      },
    ],
  },
  // Default parameters.
  {
    code: dedent`
      function f(x, y = 10) {}
      const foo = x => f(x);`,
    optionsSet: [[{ assumeTypes: { allowFixer: true } }]],
    output: dedent`
      function f(x, y = 10) {}
      const foo = f;`,
    errors: [
      {
        messageId: "generic",
        type: "ArrowFunctionExpression",
        line: 2,
        column: 13,
      },
    ],
  },
];

// Invalid test cases.
const tsInvalid: ReadonlyArray<InvalidTestCase> = [
  ...es6Invalid,
  // FunctionDeclaration.
  {
    code: dedent`
      function f(x) {}
      const foo = x => f(x);`,
    optionsSet: [[]],
    output: dedent`
      function f(x) {}
      const foo = f;`,
    errors: [
      {
        messageId: "generic",
        type: "ArrowFunctionExpression",
        line: 2,
        column: 13,
      },
    ],
  },
  // FunctionExpression.
  {
    code: dedent`
      const f = function(x) {}
      const foo = x => f(x);`,
    optionsSet: [[]],
    output: dedent`
      const f = function(x) {}
      const foo = f;`,
    errors: [
      {
        messageId: "generic",
        type: "ArrowFunctionExpression",
        line: 2,
        column: 13,
      },
    ],
  },
  // ArrowFunction.
  {
    code: dedent`
      const f = x => {}
      const foo = x => f(x);`,
    optionsSet: [[]],
    output: dedent`
      const f = x => {}
      const foo = f;`,
    errors: [
      {
        messageId: "generic",
        type: "ArrowFunctionExpression",
        line: 2,
        column: 13,
      },
    ],
  },
  // TypeAlias.
  {
    code: dedent`
      type F = (x) => {};
      const f = undefined as unknown as F;
      const foo = x => f(x);`,
    optionsSet: [[]],
    output: dedent`
      type F = (x) => {};
      const f = undefined as unknown as F;
      const foo = f;`,
    errors: [
      {
        messageId: "generic",
        type: "ArrowFunctionExpression",
        line: 3,
        column: 13,
      },
    ],
  },
  // Optional parameters.
  {
    code: dedent`
      function f(x: number, y?: number) {}
      const foo = x => f(x);`,
    optionsSet: [[]],
    output: dedent`
      function f(x: number, y?: number) {}
      const foo = f;`,
    errors: [
      {
        messageId: "generic",
        type: "ArrowFunctionExpression",
        line: 2,
        column: 13,
      },
    ],
  },
];

describeTsOnly("TypeScript", () => {
  const ruleTester = new RuleTester(typescript);
  ruleTester.run(name, rule, {
    valid: processValidTestCase(tsValid),
    invalid: processInvalidTestCase(tsInvalid),
  });
});

describe("JavaScript (es6)", () => {
  const ruleTester = new RuleTester(es6);
  ruleTester.run(name, rule, {
    valid: processValidTestCase(es6Valid),
    invalid: processInvalidTestCase(es6Invalid),
  });
});
