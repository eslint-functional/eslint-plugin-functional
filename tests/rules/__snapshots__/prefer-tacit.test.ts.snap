// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`prefer-tacit > typescript > options > checkMemberExpressions > report member expressions when enabled 1`] = `
{
  "fixed": false,
  "messages": [
    {
      "column": 1,
      "endColumn": 35,
      "endLine": 2,
      "line": 2,
      "message": "Potentially unnecessary function wrapper.",
      "messageId": "generic",
      "nodeType": "FunctionDeclaration",
      "ruleId": "prefer-tacit",
      "severity": 2,
      "suggestions": [
        {
          "desc": "Remove unnecessary function wrapper.",
          "fix": {
            "range": [
              46,
              80,
            ],
            "text": "const foo = a.b.bind(a);",
          },
          "messageId": "genericSuggestion",
        },
      ],
    },
  ],
  "output": "declare const a: { b(arg: string): string; };
function foo(x) { return a.b(x); }",
  "steps": [
    {
      "fixed": false,
      "messages": [
        {
          "column": 1,
          "endColumn": 35,
          "endLine": 2,
          "line": 2,
          "message": "Potentially unnecessary function wrapper.",
          "messageId": "generic",
          "nodeType": "FunctionDeclaration",
          "ruleId": "rule-to-test/prefer-tacit",
          "severity": 2,
          "suggestions": [
            {
              "desc": "Remove unnecessary function wrapper.",
              "fix": {
                "range": [
                  46,
                  80,
                ],
                "text": "const foo = a.b.bind(a);",
              },
              "messageId": "genericSuggestion",
            },
          ],
        },
      ],
      "output": "declare const a: { b(arg: string): string; };
function foo(x) { return a.b(x); }",
    },
  ],
}
`;

exports[`prefer-tacit > typescript > options > checkMemberExpressions > report member expressions when enabled 2`] = `
{
  "fixed": false,
  "messages": [
    {
      "column": 13,
      "endColumn": 33,
      "endLine": 1,
      "line": 1,
      "message": "Potentially unnecessary function wrapper.",
      "messageId": "generic",
      "nodeType": "ArrowFunctionExpression",
      "ruleId": "prefer-tacit",
      "severity": 2,
      "suggestions": [
        {
          "desc": "Remove unnecessary function wrapper.",
          "fix": {
            "range": [
              12,
              32,
            ],
            "text": "/a/.test.bind(/a/)",
          },
          "messageId": "genericSuggestion",
        },
      ],
    },
  ],
  "output": "[''].filter(str => /a/.test(str))",
  "steps": [
    {
      "fixed": false,
      "messages": [
        {
          "column": 13,
          "endColumn": 33,
          "endLine": 1,
          "line": 1,
          "message": "Potentially unnecessary function wrapper.",
          "messageId": "generic",
          "nodeType": "ArrowFunctionExpression",
          "ruleId": "rule-to-test/prefer-tacit",
          "severity": 2,
          "suggestions": [
            {
              "desc": "Remove unnecessary function wrapper.",
              "fix": {
                "range": [
                  12,
                  32,
                ],
                "text": "/a/.test.bind(/a/)",
              },
              "messageId": "genericSuggestion",
            },
          ],
        },
      ],
      "output": "[''].filter(str => /a/.test(str))",
    },
  ],
}
`;

exports[`prefer-tacit > typescript > reports functions that are just instantiations 1`] = `
{
  "fixed": false,
  "messages": [
    {
      "column": 1,
      "endColumn": 41,
      "endLine": 2,
      "line": 2,
      "message": "Potentially unnecessary function wrapper.",
      "messageId": "generic",
      "nodeType": "FunctionDeclaration",
      "ruleId": "prefer-tacit",
      "severity": 2,
      "suggestions": [
        {
          "desc": "Remove unnecessary function wrapper.",
          "fix": {
            "range": [
              26,
              66,
            ],
            "text": "const foo = f<number>;",
          },
          "messageId": "genericSuggestion",
        },
      ],
    },
  ],
  "output": "function f<T>(x: T): T {}
function foo(x) { return f<number>(x); }",
  "steps": [
    {
      "fixed": false,
      "messages": [
        {
          "column": 1,
          "endColumn": 41,
          "endLine": 2,
          "line": 2,
          "message": "Potentially unnecessary function wrapper.",
          "messageId": "generic",
          "nodeType": "FunctionDeclaration",
          "ruleId": "rule-to-test/prefer-tacit",
          "severity": 2,
          "suggestions": [
            {
              "desc": "Remove unnecessary function wrapper.",
              "fix": {
                "range": [
                  26,
                  66,
                ],
                "text": "const foo = f<number>;",
              },
              "messageId": "genericSuggestion",
            },
          ],
        },
      ],
      "output": "function f<T>(x: T): T {}
function foo(x) { return f<number>(x); }",
    },
  ],
}
`;

exports[`prefer-tacit > typescript > reports functions that can "safely" be changed 1`] = `
{
  "fixed": false,
  "messages": [
    {
      "column": 13,
      "endColumn": 22,
      "endLine": 2,
      "line": 2,
      "message": "Potentially unnecessary function wrapper.",
      "messageId": "generic",
      "nodeType": "ArrowFunctionExpression",
      "ruleId": "prefer-tacit",
      "severity": 2,
      "suggestions": [
        {
          "desc": "Remove unnecessary function wrapper.",
          "fix": {
            "range": [
              29,
              38,
            ],
            "text": "f",
          },
          "messageId": "genericSuggestion",
        },
      ],
    },
  ],
  "output": "function f(x) {}
const foo = x => f(x);",
  "steps": [
    {
      "fixed": false,
      "messages": [
        {
          "column": 13,
          "endColumn": 22,
          "endLine": 2,
          "line": 2,
          "message": "Potentially unnecessary function wrapper.",
          "messageId": "generic",
          "nodeType": "ArrowFunctionExpression",
          "ruleId": "rule-to-test/prefer-tacit",
          "severity": 2,
          "suggestions": [
            {
              "desc": "Remove unnecessary function wrapper.",
              "fix": {
                "range": [
                  29,
                  38,
                ],
                "text": "f",
              },
              "messageId": "genericSuggestion",
            },
          ],
        },
      ],
      "output": "function f(x) {}
const foo = x => f(x);",
    },
  ],
}
`;

exports[`prefer-tacit > typescript > reports functions that can "safely" be changed 2`] = `
{
  "fixed": false,
  "messages": [
    {
      "column": 27,
      "endColumn": 42,
      "endLine": 1,
      "line": 1,
      "message": "Potentially unnecessary function wrapper.",
      "messageId": "generic",
      "nodeType": "ArrowFunctionExpression",
      "ruleId": "prefer-tacit",
      "severity": 2,
      "suggestions": [
        {
          "desc": "Remove unnecessary function wrapper.",
          "fix": {
            "range": [
              26,
              41,
            ],
            "text": "Boolean",
          },
          "messageId": "genericSuggestion",
        },
      ],
    },
  ],
  "output": "const foo = [1, 2, 3].map(x => Boolean(x));",
  "steps": [
    {
      "fixed": false,
      "messages": [
        {
          "column": 27,
          "endColumn": 42,
          "endLine": 1,
          "line": 1,
          "message": "Potentially unnecessary function wrapper.",
          "messageId": "generic",
          "nodeType": "ArrowFunctionExpression",
          "ruleId": "rule-to-test/prefer-tacit",
          "severity": 2,
          "suggestions": [
            {
              "desc": "Remove unnecessary function wrapper.",
              "fix": {
                "range": [
                  26,
                  41,
                ],
                "text": "Boolean",
              },
              "messageId": "genericSuggestion",
            },
          ],
        },
      ],
      "output": "const foo = [1, 2, 3].map(x => Boolean(x));",
    },
  ],
}
`;
